lib/scholar/neighbors/nn_descent.ex

```diff
@@ -0,0 +1,817 @@
+  and flag.
+  """
+  defn add_neighbor(curr_graph, index0, index1, key, flag) do
+    {indices, keys, flags} = curr_graph
+    num_nodes = Nx.axis_size(indices, 1)
+
+    stop =
+      if key >= keys[[index0, 0]] do
+        Nx.u8(1)
+      else
+        Nx.u8(0)
+      end
+
+    stop = stop or Nx.any(indices[index0] == index1)
+
+    if stop do
+      curr_graph
+    else
+      {{indices, keys, flags, curr, _}, _} =
+        while {{indices, keys, flags, curr = Nx.s64(0), swap = Nx.s64(0)},
+               {index0, stop, key, flag}},
+              not stop do
+          left_child = 2 * curr + 1
+          right_child = left_child + 1
+
+          {swap, stop} =
+            cond do
+              left_child >= num_nodes ->
+                {swap, Nx.u8(1)}
+
+              right_child >= num_nodes and keys[[index0, left_child]] > key ->
+                {left_child, stop}
+
+              right_child >= num_nodes and keys[[index0, left_child]] <= key ->
+                {swap, Nx.u8(1)}
+
+              keys[[index0, left_child]] >= keys[[index0, right_child]] and
+                  keys[[index0, left_child]] > key ->
+                {left_child, stop}
+
+              keys[[index0, left_child]] >= keys[[index0, right_child]] and
+                  keys[[index0, left_child]] <= key ->
+                {swap, Nx.u8(1)}
+
+              keys[[index0, right_child]] > key ->
+                {right_child, stop}
+
+              keys[[index0, right_child]] <= key ->
+                {swap, Nx.u8(1)}
+
+              true ->
+                {swap, Nx.u8(1)}
+            end
+
+          index0_ext = Nx.new_axis(index0, 0)
+          curr_ext = Nx.new_axis(curr, 0)
+
+          indices =
+            Nx.indexed_put(
+              indices,
+              Nx.concatenate([index0_ext, curr_ext]),
+              indices[[index0, swap]]
+            )
+
+          keys =
+            Nx.indexed_put(keys, Nx.concatenate([index0_ext, curr_ext]), keys[[index0, swap]])
+
+          flags =
+            if flag != Nx.s8(-1),
+              do:
+                Nx.indexed_put(
+                  flags,
+                  Nx.concatenate([index0_ext, curr_ext]),
+                  flags[[index0, swap]]
+                ),
+              else: flags
+
+          curr = swap
+          {{indices, keys, flags, curr, swap}, {index0, stop, key, flag}}
+        end
+
+      index0_ext = Nx.new_axis(index0, 0)
+      curr_ext = Nx.new_axis(curr, 0)
+      indices = Nx.indexed_put(indices, Nx.concatenate([index0_ext, curr_ext]), index1)
+      keys = Nx.indexed_put(keys, Nx.concatenate([index0_ext, curr_ext]), key)
+
+      flags =
+        if flag != Nx.s8(-1),
+          do: Nx.indexed_put(flags, Nx.concatenate([index0_ext, curr_ext]), flag),
+          else: flags
+
+      {indices, keys, flags}
+    end
+  end
+
+  defn add_neighbor_vectorized(curr_graph, index1, key, flag, num_nodes) do
```

Comment:
Technically we could use `add_neighbor` by having vectoried `index0` with all indices. I guess it may be slightly more work, since we would use two indices to slice instead of one, etc, but perhaps it's negligible and worth simplifying the implementation.